src/Simulation/Simulators/OpenSystemsSimulator/DataModel/ArrayConverter.cs

```diff
@@ -3,249 +3,245 @@
+        // By the end, we need to have read "v", "dim", and "data".
+        int? version = null;
+        List<int>? dims = null;
+        List<Complex>? data = null;
+
+        // We require that the reader be in the right state to read an
+        // object.
+        if (reader.TokenType != JsonTokenType.StartObject)
         {
-            // By the end, we need to have read "v", "dim", and "data".
-            int? version = null;
-            List<int>? dims = null;
-            List<Complex>? data = null;
-
-            // We require that the reader be in the right state to read an
-            // object.
-            if (reader.TokenType != JsonTokenType.StartObject)
+            throw new JsonException();
+        }
+
+        while (reader.Read())
+        {
+            if (reader.TokenType == JsonTokenType.EndObject)
+            {
+                // We're at the end of the array, and can break out of the
+                // read loop.
+                break;
+            }
+
+            // If it's not the end of the object, the current token needs
+            // to be a property name.
+            if (reader.TokenType != JsonTokenType.PropertyName)
             {
                 throw new JsonException();
             }
 
-            while (reader.Read())
+            var propertyName = reader.GetString();
+
+            switch (propertyName)
             {
-                if (reader.TokenType == JsonTokenType.EndObject)
-                {
-                    // We're at the end of the array, and can break out of the
-                    // read loop.
+                case "v":
+                    reader.Read();
+                    version = reader.GetInt32();
                     break;
-                }
 
-                // If it's not the end of the object, the current token needs
-                // to be a property name.
-                if (reader.TokenType != JsonTokenType.PropertyName)
-                {
-                    throw new JsonException();
-                }
+                case "dim":
+                    dims = new List<int>();
+                    reader.Require(JsonTokenType.StartArray);
+                    while (reader.Read() && reader.TokenType != JsonTokenType.EndArray)
+                    {
+                        dims.Add(reader.GetInt32());
+                    }
+                    break;
 
-                var propertyName = reader.GetString();
+                case "data":
+                    data = new List<Complex>();
+                    reader.Require(JsonTokenType.StartArray);
+
+                    while (reader.Read() && reader.TokenType != JsonTokenType.EndArray)
+                    {
+                        // We expect an inner array at this point, so
+                        // go on and read the begin array token.
+                        reader.Require(JsonTokenType.StartArray, read: false);
```

Comment:
(I fail to (quickly) find documentation for `reader.Require()`)

Why is there in line 98 the fragment `read: false`, but in line 83 there is no? 

For now I interpret the lines 83 - 86 like this (see my comment in the code):
```c#
reader.Require(JsonTokenType.StartArray);  // Make sure that there is an array start and **consume that array start**.
while (reader.Read() ..  // Read the next token (containing the integer) after the array start.
{
    dims.Add(reader.GetInt32());   // Interpret the token as an integer.
```
But I interpret the lines 98 - 101 like this:
```c#
reader.Require(JsonTokenType.StartArray, read: false);  // Make sure that there is an array start, but **DO NOT consume that array start**.
reader.Read();   // Read the token CONTAINING THE ARRAY START.
var real = reader.GetDouble();  // Interpret the ARRAY START token as a double.
```
To summarize, I have an impression that the fragment `read: false` in line 98 results in the misinterpretation in line 101.

Reply:
`Require` is an extension method defined at https://github.com/microsoft/qsharp-runtime/blob/1334dc8cefb447e65feca66c463bcd77421bd5a2/src/Simulation/Simulators/OpenSystemsSimulator/DataModel/Extensions.cs#L66, and is used to ensure that a read succeeds with the right token.

In any case, I think your interpretation is correct, such that L103 can be safely dropped if L101 is changed to remove `read: false`.

Reply:
Your refs to the line numbers seem shifted but looking at the definition of `Require()` I understood what you meant.
Consider renaming the param `read` to `preRead` or `readBefore` or similar. Otherwise I interpreted it as a `postRead` or `readAfter`.

Reply:
Done.